File: meaty_pythons/src/resolvers.py
import os
import configparser
import logging

from ariadne import (
    QueryType,
    MutationType,
    gql,
    make_executable_schema,
)
from ariadne.asgi import GraphQL

from redis import Redis

from fastapi import FastAPI
from fastapi.staticfiles import StaticFiles

from starlette.middleware import Middleware
from starlette.middleware.cors import CORSMiddleware

from gql import Client, gql as gql_client
from gql.transport.aiohttp import AIOHTTPTransport


# Load configuration from config.ini
config = configparser.ConfigParser()
config.read("/app/config.ini")

logger = logging.getLogger(__name__)

# Replace hardcoded Redis server and port values with values from config.ini
redis_client = Redis(
    host=config.get("REDIS", "HOST"), port=config.get("REDIS", "PORT"), db=0
)

query = QueryType()
mutation = MutationType()

type_defs = gql(
    """
    type InventoryItem {
        id: ID!
        stockLevel: Int!
        price: Float!
    }

    type Vendor {
        title: String!
        description: String!
        icon: String!
        inventory: [InventoryItem!]!
    }

    enum PurchaseResult {
        SUCCESS
        INSUFFICIENT_STOCK
        ITEM_NOT_FOUND
    }

    type Query {
        vendor: Vendor!
        item(id: ID!): InventoryItem
    }

    type Mutation {
        purchase(id: ID!, quantity: Int!): PurchaseResult!
    }
"""
)


@query.field("vendor")
def resolve_vendor(*_):
    inventory = []
    for key in redis_client.keys():
        item_data = redis_client.hgetall(key)
        if item_data:
            inventory.append({k.decode(): v.decode() for k, v in item_data.items()})
        else:
            logger.error(f"Invalid data for key: {key}")

    vendor = {
        "title": "Meaty Pythons",
        "description": "We sell the best limbless reptilian meat in town!",
        "icon": "http://localhost:8082/static/python.png",
        "inventory": inventory,
    }

    return vendor


@query.field("item")
def resolve_item(*_, id):
    item_data = redis_client.hgetall(id)
    if item_data:
        item_data = {k.decode(): v.decode() for k, v in item_data.items()}
        item_data["stockLevel"] = int(item_data["stockLevel"])
        item_data["price"] = float(item_data["price"])
        return item_data
    return None


@mutation.field("purchase")
def resolve_purchase(*_, id, quantity):
    item_data = redis_client.hgetall(id)
    if item_data:
        item_data = {k.decode(): v.decode() for k, v in item_data.items()}
        if int(item_data["stockLevel"]) < quantity:
            return "INSUFFICIENT_STOCK"
        else:
            item_data["stockLevel"] = int(item_data["stockLevel"]) - quantity
            redis_client.hmset(id, item_data)
            return "SUCCESS"
    else:
        return "ITEM_NOT_FOUND"


def register_self(
    broker_url: str, vendor_id: str, url: str, title: str, description: str, icon: str
):
    logger.info(f"Connecting to broker at: {broker_url}")
    print(f"Connecting to broker at: {broker_url}")
    transport = AIOHTTPTransport(url=broker_url)
    client = Client(transport=transport, fetch_schema_from_transport=True)

    query = gql_client(
        """#graphql
            mutation Mutation(
                $vendorId: ID!
                $url: String!
                $title: String!
                $description: String!
                $icon: String!
                ) {
                    registerVendor(
                        vendorId: $vendorId
                        url: $url
                        title: $title
                        description: $description
                        icon: $icon
                    )
            }
        """
    )

    result = client.execute(
        query,
        variable_values={
            "vendorId": vendor_id,
            "url": url,
            "title": title,
            "description": description,
            "icon": icon,
        },
    )

    logger.info(f"Connection to broker: {'SUCCESSFUL' if result else 'FAILED'}!")
    print(f"Connection to broker: {'SUCCESSFUL' if result else 'FAILED'}!")


#
schema = make_executable_schema(type_defs, query, mutation)

# Instantiate FastAPI with custom middleware for CORS handling
app = FastAPI(middleware=[Middleware(CORSMiddleware, allow_origins=["*"])])

# Mount the GraphQL app at /query
app.add_route("/", GraphQL(schema, debug=True))

# Mount static files at /static
app.mount("/static", StaticFiles(directory="/app/static"), name="static")

if __name__ == "__main__":
    import uvicorn

    logger.info("Registering self with broker")
    broker_url = os.environ.get("BROKER_URL")

    register_self(
        broker_url,
        "meaty-pythons",
        "http://meaty-pythons:8082/",
        "Meaty Pythons",
        "We sell the best limbless reptilian meat in town!",
        "http://meaty-pythons:8082/static/python.png",
    )
    logger.info("Starting GraphQL server")
    uvicorn.run(app, host="0.0.0.0", port=8082)

File: fruitful_javascripts/src/resolvers.js
import express from "express";
import { ApolloServer, gql } from "apollo-server-express";
import request, { gql as req_gql } from "graphql-request";
import sqlite3 from "sqlite3";

// Create a new SQLite database connection
const db = new sqlite3.Database(
  "/data/fruitful.db", // <-- update this line
  sqlite3.OPEN_READWRITE,
  err => {
    if (err) {
      console.error(err.message);
    }
    console.log("Connected to the fruitful.db SQLite database.");
  },
);

const typeDefs = gql`
  type InventoryItem {
    id: ID!
    stockLevel: Int!
    price: Float!
  }

  type Vendor {
    title: String!
    description: String!
    icon: String!
    inventory: [InventoryItem!]!
  }

  enum PurchaseResult {
    SUCCESS
    INSUFFICIENT_STOCK
    ITEM_NOT_FOUND
  }

  type Query {
    vendor: Vendor!
    item(id: ID!): InventoryItem
  }

  type Mutation {
    purchase(id: ID!, quantity: Int!): PurchaseResult!
  }
`;

// Define resolvers
const resolvers = {
  Query: {
    vendor: (_, __, { db }) => {
      return {
        title: "Fruitful JavaScripts",
        description: "We sell the best fruits in town!",
        icon: "http://localhost:8084/static/fruits.png",
        inventory: new Promise((resolve, reject) => {
          // Fetch inventory data from the database
          db.all("SELECT * FROM inventory", (err, rows) => {
            if (err) reject(err);
            else
              resolve(
                rows.map(row => ({
                  ...row,
                  stockLevel: row.stock_level,
                })),
              );
          });
        }),
      };
    },
    item: (_, { id }, { db }) => {
      return new Promise((resolve, reject) => {
        // Fetch item data from the database based on ID
        db.get("SELECT * FROM inventory WHERE id = ?", [id], (err, row) => {
          if (err) {
            console.log("Error fetching item from database");
            reject(err);
          } else if (Boolean(row)) {
            resolve({
              ...row,
              stockLevel: row.stock_level,
            });
          } else {
            console.log("Item not found in database");
            reject("Item not found");
          }
        });
      });
    },
  },
  Mutation: {
    purchase: (_, { id, quantity }, { db }) => {
      return new Promise((resolve, reject) => {
        db.serialize(() => {
          // Fetch the stock level of the item from the database
          db.get(
            "SELECT stock_level FROM inventory WHERE id = ?",
            [id],
            (err, row) => {
              if (err) {
                reject(err);
              } else if (!row) {
                resolve("ITEM_NOT_FOUND");
              } else if (row.stock_level < quantity) {
                resolve("INSUFFICIENT_STOCK");
              } else {
                // Update the stock level in the database
                db.run(
                  "UPDATE inventory SET stock_level = stock_level - ? WHERE id = ?",
                  [quantity, id],
                  function (err) {
                    if (err) {
                      reject(err);
                    } else {
                      if (this.changes > 0) {
                        resolve("SUCCESS");
                      } else {
                        resolve("ITEM_NOT_FOUND");
                      }
                    }
                  },
                );
              }
            },
          );
        });
      });
    },
  },
};

const startApolloServer = async () => {
  const apolloServer = new ApolloServer({
    typeDefs,
    resolvers,
    context: { db },
  });

  await apolloServer.start();

  const app = express();
  apolloServer.applyMiddleware({ app, path: "/" });
  app.use("/static", express.static("../static"));

  const port = process.env.RESOLVERS_PORT || 8084;
  app.listen(port, () => {
    console.log(`Server is running on http://localhost:${port}`);
  });
};

const registerService = async (brokerUrl, params) => {
  await new Promise(resolve => setTimeout(resolve, 5000));
  const mutation = req_gql`#graphql
    mutation Mutation(
        $vendorId: ID!
        $url: String!
        $title: String!
        $description: String!
        $icon: String!
        ) {
            registerVendor(
                vendorId: $vendorId
                url: $url
                title: $title
                description: $description
                icon: $icon
            )
    }
  `;
  console.log(`Registering service with params: ${JSON.stringify(params)}`);
  const result = await request(brokerUrl, mutation, params);
  console.log(`Registered service with result: ${result.registerVendor}`);
};

startApolloServer().catch(err => {
  console.error("Error starting Apollo Server:", err);
});

const brokerUrl = process.env.BROKER_URL;

// Register the vendor service with the broker
registerService(brokerUrl, {
  vendorId: "fruitful-javascripts",
  url: "http://fruitful-javascripts:8084",
  title: "Fruitful JavaScripts",
  description: "We sell the best fruits in town!",
  icon: "http://localhost:8084/static/fruits.png",
});
